File: example/crc32/crc.go
package main

import "fmt"

var crc_table [256]uint32

func crc_init_1(n uint32) uint32 {
    for i := uint32(0); i < 8; i++ {
        if (n & 1) != 0 {
            n = (n >> 1) ^ 0xedb88320
        } else {
            n >>= 1
        }
    }
    return n
}


func init() {
    for n := uint32(0); n < 256; n++ {
        crc_table[n] = crc_init_1(n)
    }
}

func crc_add_byte(crc uint32, b uint8) uint32 {
    return (crc >> 8) ^ crc_table[uint8(crc) ^ b]
}

func crc_n_bytes(b uint8, n uintptr) uint32 {
    crc := uint32(0xffffffff)
    for i := uintptr(0); i < n; i++ {
        crc = crc_add_byte(crc, b)
    }
    return crc;
}

func main() {
    fmt.Println(crc_n_bytes('t', 100 * 1000 * 1000))
}

File: example/crc32c/_crc.c
#include <stdint.h>
#include <stdio.h>

static uint32_t crc_make_table_element(uint32_t n) {
    uint32_t i;
    for (i = 0; i < 8; i++) {
        if (n & 1) {
            n = (n >> 1) ^ 0xedb88320;
        } else {
            n >>= 1;
        }
    }
    return n;
}

uint32_t crc_table[256];

static void crc_make_table(void) {
    uint32_t n;
    for (n = 0; n < 256; n++) {
        crc_table[n] = crc_make_table_element(n);
    }
}

#if 0 /* enable to use an assembly implementation of crc_add_byte() */
uint32_t crc_add_byte(uint32_t crc, uint8_t byte);
#else
uint32_t crc_add_byte(uint32_t crc, uint8_t byte) {
    return (crc >> 8) ^ crc_table[(crc & 0xff) ^ byte];
}
#endif /* 0 */

uint32_t crc_n_bytes(uint8_t byte, size_t n) {
    uint32_t crc = 0xffffffff;
    size_t i;
    for (i = 0; i < n; i++) {
        crc = crc_add_byte(crc, byte);
    }
    return crc;
}

int main(void) {
    crc_make_table();
    printf("%u\n", (unsigned)crc_n_bytes('t', 1000 * 1000 * 1000));
    return 0;
}
